moveit_core/trajectory_processing/src/ruckig_traj_smoothing.cpp

```diff
@@ -0,0 +1,169 @@
+constexpr double DEFAULT_MAX_VELOCITY = 5;       // rad/s
+constexpr double DEFAULT_MAX_ACCELERATION = 10;  // rad/s^2
+constexpr double DEFAULT_MAX_JERK = 20;          // rad/s^3
+
+bool RuckigSmoothing::computeTimeStamps(robot_trajectory::RobotTrajectory& trajectory,
+                                        const double max_velocity_scaling_factor,
+                                        const double max_acceleration_scaling_factor) const
+{
+  const moveit::core::JointModelGroup* group = trajectory.getGroup();
+  if (!group)
+  {
+    RCLCPP_ERROR(LOGGER, "It looks like the planner did not set the group the plan was computed for");
+    return false;
+  }
+
+  size_t num_waypoints = trajectory.getWayPointCount();
+  if (num_waypoints < 2)
+  {
+    RCLCPP_WARN_STREAM(LOGGER, "Trajectory does not have enough points to smooth with Ruckig");
+    return false;
+  }
+
+  const size_t num_dof = group->getVariableCount();
+
+  // This lib does not actually work properly when angles wrap around, so we need to unwind the path first
+  trajectory.unwind();
+
+  // Instantiate the smoother
+  double timestep = trajectory.getAverageSegmentDuration();
+  ruckig::Ruckig<0> ruckig{ num_dof, timestep };
+  ruckig::InputParameter<0> ruckig_input{ num_dof };
+  ruckig::OutputParameter<0> ruckig_output{ num_dof };
+
+  // Initialize the smoother
+  //  ruckig_input.synchronization = ruckig::Synchronization::Time;
+  //  ruckig_input.interface = ruckig::Interface::Velocity;
+  const std::vector<int>& idx = group->getVariableIndexList();
+  std::vector<double> current_positions_vector(num_dof);
+  std::vector<double> current_velocities_vector(num_dof);
+  std::vector<double> current_accelerations_vector(num_dof);
+  moveit::core::RobotStatePtr waypoint = trajectory.getFirstWayPointPtr();
+  for (size_t i = 0; i < num_dof; ++i)
+  {
+    current_positions_vector.at(i) = waypoint->getVariablePosition(idx[i]);
+    current_velocities_vector.at(i) = waypoint->getVariableVelocity(idx[i]);
+    current_accelerations_vector.at(i) = waypoint->getVariableAcceleration(idx[i]);
+  }
+  std::copy_n(current_positions_vector.begin(), num_dof, ruckig_input.current_position.begin());
+  std::copy_n(current_velocities_vector.begin(), num_dof, ruckig_input.current_velocity.begin());
+  std::copy_n(current_accelerations_vector.begin(), num_dof, ruckig_input.current_acceleration.begin());
+  // Initialize output data struct
+  ruckig_output.new_position = ruckig_input.current_position;
+  ruckig_output.new_velocity = ruckig_input.current_velocity;
+  ruckig_output.new_acceleration = ruckig_input.current_acceleration;
+
+  // Kinematic limits (vel/accel/jerk)
+  const std::vector<std::string>& vars = group->getVariableNames();
+  const moveit::core::RobotModel& rmodel = group->getParentModel();
+  for (size_t i = 0; i < num_dof; ++i)
+  {
+    // TODO(andyz): read this from the joint group if/when jerk becomes supported
```

Comment:
can you clarify in the comment... when jerk becomes supported by the moveit trajectory_processing API? or by something that ruckig changes?

Reply:
We need to get jerk added and parsed from joint_limits.yaml

Reply:
> We need to get jerk added and parsed from joint_limits.yaml

If we get to really supporting it, this sounds like a great idea!